internal/docstore/docstore.go

```diff
@@ -180,21 +197,71 @@ func (l *ActionList) Update(doc Document, mods Mods) *ActionList {
 // See ActionList.Update.
 type Mods map[FieldPath]interface{}
 
-// Do executes the action list. If all the actions executed successfully, Do returns
-// (number of actions, nil). If any failed, Do returns the number of successful
-// actions and an error. In general there is no way to know which actions succeeded,
-// but the error will contain as much information as possible about the failures.
-func (l *ActionList) Do(ctx context.Context) (int, error) {
+// An ActionListError is returned by ActionList.Do. It contains all the errors
+// encountered while executing the ActionList, and the positions of the corresponding
+// actions.
+type ActionListError []struct {
+	Index int
+	Err   error
+}
+
+// TODO(jba): use xerrors formatting.
+
+func (e ActionListError) Error() string {
+	var s []string
+	for _, x := range e {
+		s = append(s, fmt.Sprintf("at %d: %v", x.Index, x.Err))
+	}
+	return strings.Join(s, "; ")
+}
+
+// Unwrap returns the error in e, if there is exactly one. If there is more than one
+// error, Unwrap returns nil, since there is no way to determine which should be
+// returned.
+func (e ActionListError) Unwrap() error {
+	if len(e) == 1 {
+		return e[0].Err
+	}
+	// Return nil when e is nil, or has more than one error.
+	// When there are multiple errors, it doesn't make sense to return any of them.
+	return nil
+}
+
+// Do executes the action list. If any action fails, the returned error will be an
+// ActionListError that contains the position in the ActionList of each failed
+// action.
+//
+// In ordered mode (when the Unordered method was not called on the list), execution
+// will stop after the first action that fails.
+//
+// In unordered mode, all the actions will be executed. In either mode, as a special
+// case, none of the actions will be executed if any is invalid (for example, a Put
+// whose document is missing its key field).
+func (l *ActionList) Do(ctx context.Context) error {
```

Comment:
If this function always returns `ActionListError`, maybe state that more explicitly?

Reply:
Clarified.